### ISSUE TITLE
Make SendPort and ReceivePort generic

### ISSUE THREAD

**Issue by Levi-Lesches** [dart-lang/language#3432]
Currently, [`SendPort.send`](https://api.dart.dev/stable/3.1.5/dart-isolate/SendPort/send.html) accepts any `Object?`, and `ReceivePort` implements `Stream<dynamic>`. I see a nice opportunity to add some type safety here: 

```dart
class SendPort<T> {
  void send(T object);
}

class ReceivePort<T> implements Stream<T> {
  final SendPort<T> sendPort;
}

// Safer code: 
void main() {
  final numberPort = ReceivePort<num>();
  Isolate.spawn(runIsolate, numberPort.sendPort);
  // implied: Isolate.spawn<SendPort<num>>(...)
}

void runIsolate(SendPort<num> port) {
  port.send(1);  // OK
  port.send(3.14);  // OK
  port.send("Hello");  // Error
}
```

Or even, with pattern matching: 
```dart
sealed class IsolateMessage { } 
class RefreshPage extends IsolateMessage { }
class FetchData extends IsolateMessage { }

final port = ReceivePort<IsolateMessage>()
  ..listen((message) => switch(message) {
    RefreshPage() => print("Refreshing..."),
    FetchData() => print("Fetching..."),
  });
```

**Comment by abitofevrything** [comment]
Duplicate of https://github.com/dart-lang/sdk/issues/23059

**Comment by lrhn** [comment]
Good find. And the issues mentioned on that issue still apply.

A solution could be "TypeReveivePort"/"TypedSendPort" types which cannot be sent to isolated that are not in the same isolate group.
They would likely _not_ implement `ReveivePort` or `SendPort`, because then upcasting them would lose that information. It's a new _like_ of port.

It's very likely easier to just create wrappers yourself, or just do `receivePort.cast<MyMessage>()`.
(If native code gets hold of a send port, they can still send any value, so done receiver side checking is necessary anyway.)

**Comment by Levi-Lesches** [comment]
Right, I think limiting this to the case of isolates in the same group would still be useful. Although I agree that making our own wrappers is easy, they can't implement the original `SendPort` or `ReceivePort`, because then they'll be forced to inherit/define the unsafe methods, so they won't be compatible with any of the existing APIs.
